### PR TITLE
Make contract metrics good for the front end

### PR DESCRIPTION
**Pull request by IanPhilips**
updates contract metrics onCreateBet, create and backfill contractMetrics with denormalized user data

### CODE REVIEW

**Comment by mqp** [comment]
I think this is probably a good change for right now, although I don't know if it's going to scale well for very long. But why are you still recomputing them in `updateUserMetrics`?

File: functions/src/on-create-bet.ts [hunk 1]
```diff
@@ -305,3 +307,23 @@ const notifyFills = async (
 const currentDateBettingStreakResetTime = () => {
   return new Date().setUTCHours(BETTING_STREAK_RESET_HOUR, 0, 0, 0)
 }
+
+const updateContractMetrics = async (contract: Contract, user: User) => {
+  const betSnap = await firestore
+    .collection(`contracts/${contract.id}/bets`)
+    .where('userId', '==', user.id)
+    .get()
```
Comment (mqp) [inline functions/src/on-create-bet.ts:315]: Does this include the current bet that triggered the function? It does, doesn't it?

File: functions/src/scripts/backfill-contract-metrics-user-data.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      await Promise.all(
+        contractMetricsWithoutUserData.map(async (cm) => {
+          await firestore
+            .doc(`users/${u.id}/contract-metrics/${cm.contractId}`)
+            .set({
+              ...cm,
+              userId: user.id,
+              userUsername: user.username,
+              userName: user.name,
+              userAvatarUrl: user.avatarUrl,
```
Comment (mqp) [inline functions/src/scripts/backfill-contract-metrics-user-data.ts:45]: This needs to be fixed up when user info changes.

File: web/hooks/use-contract-metrics.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+export function listenForBinaryContractMetricsOnContract(
+  contractId: string,
+  count: number,
+  outcome: 'YES' | 'NO',
+  setMetrics: (metrics: ContractMetric[]) => void
+) {
+  const yesQuery = query(
+    collectionGroup(db, 'contract-metrics'),
+    where('contractId', '==', contractId),
+    where(outcome === 'YES' ? 'hasYesShares' : 'hasNoShares', '==', true),
```
Comment (mqp) [inline web/hooks/use-contract-metrics.ts:56]: Are these fields useful? Could the query equally well just be checking for a positive number of shares? Just curious.

**Comment by mqp** [comment]
I'll be waiting on this PR with the presumption that it will cause `lastBetTime` to get updated immediately upon bets, and then I'll likely be getting rid of all the precomputed `uniqueBettor` fields (at least `uniqueBettorIds` which is very heavy) by querying `lastBetTime`.

File: functions/src/on-create-bet.ts (hunk 1 above)
Comment (IanPhilips) [reply functions/src/on-create-bet.ts:315]: I think it must, too!

File: functions/src/scripts/backfill-contract-metrics-user-data.ts (hunk 2 above)
Comment (IanPhilips) [reply functions/src/scripts/backfill-contract-metrics-user-data.ts:45]: Yeah :(

File: web/hooks/use-contract-metrics.ts (hunk 3 above)
Comment (IanPhilips) [reply web/hooks/use-contract-metrics.ts:56]: I will try that! If that works then that would solve the FR case

**Comment by IanPhilips** [comment]
> I think this is probably a good change for right now, although I don't know if it's going to scale well for very long. But why are you still recomputing them in `updateUserMetrics`?

Just haven't gotten around to thinking about that yet, but thanks for the tip! I'll remove it there.

File: web/hooks/use-contract-metrics.ts (hunk 3 above)
Comment (IanPhilips) [reply web/hooks/use-contract-metrics.ts]: Okay, so there's a clear path for non-binary queries now but the above queries are useful bc they cut down on the filtering needed by the client

**Pull request merged**